docs/auth.md

```diff
@@ -0,0 +1,96 @@
+- APIs for logging in users or restricting content
+
+The authentication system in Soul aims to be very generic and doesn't provide some features commonly found in web authentication systems, such as:
+
+- Password strength checking
+- Throttling of login attempts
+- Authentication against third-parties (OAuth, for example)
+- Object-level permissions
+
+#### Users
+
+User objects are the core of the authentication system. They typically represent the people interacting with your site and are used to enable things like restricting access, registering user profiles, associating content with creators, etc. Only one class of user exists in Soul's authentication framework, i.e., 'superusers' or types of users are just user objects with special attributes set, not different classes of user objects.
+
+The primary attributes of the default user are:
+
+- username
+- password
+- email
+- first_name
+- last_name
+
+Note that when Soul boots up, it looks for a table called "\_users" which holds the User objects mentioned above.
+
+Creating superusers:
+
+Create superusers using the createsuperuser command:
+
+```
+$ node src/server.js createsuperuser --username=joe --password=strongstring
+```
+
+#### Permissions
+
+Soul comes with a built-in permissions system. It provides a way to assign permissions to specific users and groups of users.
+
+Soul uses permissions as follows:
+
+- Access to view objects is limited to users with the "read" permission for that type of object.
+- Access to add an object is limited to users with the "create" permission for that type of object.
+- Access to change an object is limited to users with the "update" permission for that type of object.
+- Access to delete an object is limited to users with the "delete" permission for that type of object.
+
+Default Permissions:
+
+By default, Soul ensures that four default permissions – create, update, delete, and read – are created for each table defined in one of your databases.
+
+| table_name | can_create | can_read | can_update | can_delete |
+| ---------- | ---------- | -------- | ---------- | ---------- |
+| albums     | false      | true     | true       | false      |
+
+In this scenario, calling POST (create) and GET (read) requests to /tables/albums/rows is allowed by any user (even anonymous), while PUT (update) and DELETE (delete, of course) requests are not allowed.
+
+When Soul boots up, it checks for the existence of a table called \_default_permissions.
+
+#### Roles
+
+Roles are a generic way of categorizing users so you can apply permissions or some other label to those users. A user can belong to any number of roles.
+
+A user with a role automatically has the permissions granted to that role. For example, if the role "editor" has the permission can_update_posts, any user with that role will have that permission.
+
+Beyond permissions, roles are a convenient way to categorize users and give them some label or extended functionality.
+
```

Comment:
As before - I guess there will be `roles` table. Could you give an example of what it looks like?

Reply:
Sure, roles have a "name" field that provides a user-friendly name for the role. They also have a many-to-many relationship with the permissions table, which means a role can have multiple permissions assigned to it. 
Also, the permissions table consists of a "table name" field and an enum field that represents the specific operations allowed, e.g. create, read, update, and delete (CRUD). 
Note that when a new table is created, the system should check whether all four permissions for that table (create, read, update, and delete) already exist. If any of these permissions are missing, they should be created accordingly.

Reply:
Ok. So, for the many-to-many relationship we'll need another table (a _bridging table_), right?

Reply:
Yeah, exactly.

Reply:
Note: the above strategy has changed.  We don't have many-to-many between roles and permissions.  We have multiple permissions rows per role (one per table that the role can see).